### PR TITLE
Add window module to pycbc

### PR DESCRIPTION
**Pull request by cdcapano**
This patch adds `pycbc/window.py`, which is module for generating and applying windows to data. At the moment, the main class this module introduces, `TimeDomainWindow`, isn't used by anything else. I'll be filing a separate pull request to use this class in the waveform generator for PE (it's needed for testing the area theorem). The main effect this patch has on the rest of the code is to move `taper_timeseries` from `waveform/utils.py` to `window.py`, and have all codes that use `taper_timeseries` point to the new location.

I figure the `TimeDomainWindow` class could have broader uses than just for waveform generation in the future, which is why I stuck it in the top level. The class allows you to use any window function in `scipy.signal` to use for the left and right taper of the window, along with LALSimulation's taper function. You can use different functions and taper durations for each side, and make the window arbitrarily large. Right now the window it creates tapers up from 0 to 1, then back down from 1 to 0. It would be straightforward to use this class to also create gates (you'd just need to swap left and right), so might be useful if anyone wants to investigate using something other than a Tukey window for gating.

### CODE REVIEW

File: pycbc/window.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+
+    See Also
+    --------
+    scipy.signal.get_window : function for generating windows
+    laltaper_timeseries : function that only applies the lal taper
+    """
+    def __init__(self, left_taper=None, right_taper=None,
+                 left_taper_duration=None, right_taper_duration=None,
+                 window_whitened=False, psds=None):
+        if int(window_whitened) not in [UNWHITENED, WHITENED, OVERWHITENED]:
```
Comment (titodalcanton) [inline pycbc/window.py]: Personally I would use a different approach for selecting whitening: an argument `whiten` which can be `None` (no whitening) or the strings `'whiten'` or `'overwhiten'`. So there is no need to define and import constants, and the code is more straightforward.

File: pycbc/window.py [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+        # check times
+        if left_time is not None:
+            # check that a time wasn't specified for lal tapers
+            if self.left_taper == 'lal':
+                raise ValueError("lal tapers do not require a time")
+            # check that we won't be left with all zeros
+            if isinstance(h, FrequencySeries):
+                end_time = 1./h.delta_f
+            else:
+                end_time = len(h)*h.delta_t
```
Comment (titodalcanton) [inline pycbc/window.py]: You can simply use `h.duration` here.

File: pycbc/window.py [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+                break_indx = -int(break_time/ht.delta_t)
+                ht.roll(break_indx)
+                rollback = -break_indx
+                # set break time to 0, so don't roll again
+                break_time = 0
+            tmeth = ''
+            if self.left_taper == 'lal':
+                tmeth = 'start'
+            if self.right_taper == 'lal':
+                tmeth = ''.join([tmeth, 'end'])
```
Comment (titodalcanton) [inline pycbc/window.py]: `tmeth += 'end'`?

**Comment by titodalcanton** [comment]
I have another more general comment (which I also wrote in the review but somehow got lost when I submitted it).

It is not clear to me why a generic `TimeDomainWindow` class needs to implement whitening. It seems it is being a bit too specific for the application you have in mind. The same comment applies to the fact that the PSD can be given as a dictionary indexed by an ifo. I can see the usefulness of the approach for high-perfomance simultaneous windowing in the context of multidetector data processing, but then is it really worth splitting this class to its own module? What other application could benefit from it?

**Comment by titodalcanton** [comment]
@cdcapano is this still active?

**Comment by cdcapano** [comment]
@titodalcanton Yes, this is still active. I haven't had a chance to address your comments yet as we've been making changes to this for the area theorem. Will get back to updating this patch soon.

**Comment by spxiwh** [comment]
Any updates on this? If there's nothing new the next time I look over old pull requests then I'll close this ... It's easy to open a new pull request later.

**Comment by duncan-brown** [comment]
@cdcapano I actually need this functions for something else that I'm working on. Are you still working on this? I'm happy to help fix the things that @titodalcanton is concerned about, as it will be easier to me to fix this than start from scratch.

**Comment by cdcapano** [comment]
@duncan-brown I'm working on it in a slightly tangential way. We've been using this for the area theorem test; this was going to be the first in a series of patches. I've been slow to update this patch because we had some recent results that seemed to suggest that the windowing we were doing wasn't working. However, now we have newer results that suggest it is working, so that might be moot.

Anyway, yeah, if you'd like to address @titodalcanton's comments, please feel free. I haven't had a chance to clean this up more for master yet, and probably won't for awhile. Out of curiosity, what do you need this for?

**Comment by spxiwh** [comment]
Closing this as it's seen no activity in 6 months. Please re-open if this becomes ready to merge.

**Pull request closed without merging**